### PR TITLE
[#11208] Distribute points among recipients: show average excluding self

### LINKED ISSUE #11208: Distribute points among recipients: show average excluding self
Requested by a user:

Context: an instructor is viewing the results of a 'distribute points among recipients' type question

image

Suggestion: Also show 'average excluding self' column

### PR DESCRIPTION
**Pull request by Jellybeano**
Fixes #11208

**Outline of Solution**

Added column that displays the Average Excluding Self for questions that require distributing points among recipients.

image

### CODE REVIEW

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts [hunk 1]
```diff
@@ -55,6 +57,19 @@ export class ConstsumRecipientsQuestionStatisticsCalculation
       this.totalPointsPerOption[recipient] = sum;
       this.averagePointsPerOption[recipient] = +(answers.length === 0 ? 0 : sum / answers.length).toFixed(2);
     }
+
+    let totalSum: number = 0;
+    let totalStudents: number = 0;
+    for (const recipient of Object.keys(this.pointsPerOption)) {
+      totalSum += this.totalPointsPerOption[recipient];
+      totalStudents += 1;
+    }
```
Comment (halfwhole) [inline src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts:66]: It looks like `totalSum` is the same as `sum` as defined previously, and `totalStudents` can be found easily through `answers.length` too (similar to how the average points per option was calculated above)

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts [hunk 2]
```diff
@@ -55,6 +57,19 @@ export class ConstsumRecipientsQuestionStatisticsCalculation
...
+
+    for (const recipient of Object.keys(this.pointsPerOption)) {
+      this.averagePointsExcludingSelf[recipient] =
+          +(totalStudents === 0 ? 0 : (totalSum - this.totalPointsPerOption[recipient]) / totalStudents).toFixed(2);
+    }
```
Comment (halfwhole) [inline src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts:71]: Not really sure what is being attempted here! Our goal is to find out who the 'self' is, and find a way to exclude their score from the total `sum`. This can be done by possibly comparing the response's recipient to the response's giver, an example of this can be found in say `rankPerOptionExcludeSelf` of `rank-recipients-question-statistics-calculation.ts`

**Review by halfwhole — CHANGES_REQUESTED** [review]
Thank you for contributing!

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts (hunk 1 above)
Comment (Jellybeano) [reply src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts:66]: thanks for helping look through the code!

it would seem that the `sum` might have caused some misunderstanding here, as `sum` refers to the sum of a student's points received (which will be reflected under the **Total Points** column)

`totalSum` on the other hand calculates the sum of all the student's total scores (in the above screenshot example, that would be 800)

on the same note, `answers.length` always returns 1 for each student as each of them has only one instance of receiving points in the above example

however I did notice that this extra loop was unnecessary and could be omitted, so thanks so much for pointing this out!

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts (hunk 2 above)
Comment (Jellybeano) [reply src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts:71]: Apologies for not providing a better explanation in the original commit! What the above code snippet does is that it runs through each recipient, and takes the total of all student scores (`totalSum`), substracts the score of the recipient (`this.totalPointsPerOption[recipient]`) and divides it by all students (`totalStudents`) to obtain the average. 

while `rankPerOption` and `rankPerOptionExcludeSelf` can be computed simultaneously due to their similar nature, `Average` and `AverageExceptSelf` are slightly different as one is the average of one student's scores while the other is the average of all students' scores excluding one student...unless I am understanding something wrong here. Much apologies if I am!

it is because of this reason that I put the two computations in separate loops, the computation for `AverageExceptSelf` only takes place after the total of all students scores is computed.

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts (hunk 1 above)
Comment (halfwhole) [reply src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts]: You're right, I see the intent behind introducing `totalSum` and `totalStudents` now. However, why should we need to find the total sum of scores across students, or the total number of students, at all?

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts (hunk 2 above)
Comment (halfwhole) [reply src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts:71]: > `Average` and `AverageExceptSelf` are slightly different as one is the average of one student's scores while the other is the average of all students' scores excluding one student

From what I understand, 'average excluding self' might have a different meaning: it is the average of all scores received by _one_ recipient, excluding the recipient's own score (if applicable).

Suppose recipient Gene receives two scores: 100 from Gene (a self-review) and 50 from Charlie. Then the average points received should be 75, and the average points received excluding self should be 50, as it discounts the 100 from self-review.

But under your current scheme, the average excluding self returns 131.25:

Screenshot 2021-06-28 at 11 32 20 PM

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts [hunk 3]
```diff
@@ -19,6 +19,9 @@ export class ConstsumRecipientsQuestionStatisticsCalculation
   pointsPerOption: Record<string, number[]> = {};
   totalPointsPerOption: Record<string, number> = {};
   averagePointsPerOption: Record<string, number> = {};
+  averagePointsExcludingSelf: Record<string, number> = {};
+  totalSum: number = 0;
+  totalStudents: number = 0;
```
Comment (halfwhole) [inline src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts:24]: (On a sidenote, it might not be a good idea to have variables like these used temporarily for computation only declared as class properties! Variables should try to live in the smallest scope possible.)

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts (hunk 2 above)
Comment (Jellybeano) [reply src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts:74]: Ah i understand what went wrong now! My bad, it seems like I misunderstood the meaning behind `Average Excluding Self`. That would explain the slight confusion: I have amended the code accordingly! Thanks so much for helping to clarify things!

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts (hunk 3 above)
Comment (Jellybeano) [reply src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts]: I see! Thank you for the guidance!

File: src/web/app/components/question-types/question-statistics/constsum-recipients-question-statistics.component.ts [hunk 4]
```diff
@@ -44,6 +44,7 @@ export class ConstsumRecipientsQuestionStatisticsComponent extends ConstsumRecip
       { header: 'Points Received' },
       { header: 'Total Points', sortBy: SortBy.CONSTSUM_RECIPIENTS_POINTS },
       { header: 'Average Points', sortBy: SortBy.CONSTSUM_RECIPIENTS_POINTS },
+      { header: 'Average Excluding Self', sortBy: SortBy.CONSTSUM_RECIPIENTS_POINTS },
```
[the remark this answers is not in the document]
Comment (Jellybeano) [reply src/web/app/components/question-types/question-statistics/constsum-recipients-question-statistics.component.ts:47]: oh that is very true, thanks for pointing that out!

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts [hunk 5]
```diff
@@ -17,8 +17,10 @@ export class ConstsumRecipientsQuestionStatisticsCalculation
   emailToTeamName: Record<string, string> = {};
   emailToName: Record<string, string> = {};
   pointsPerOption: Record<string, number[]> = {};
+  pointsPerOptionToSelf: Record<string, number[]> = {};
```
[the remark this answers is not in the document]
Comment (Jellybeano) [reply src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts]: it would indeed make sense for only required values to be stored, since this value is just an intermediate

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts [hunk 6]
```diff
@@ -40,6 +44,11 @@ export class ConstsumRecipientsQuestionStatisticsCalculation
       this.pointsPerOption[identifier] = this.pointsPerOption[identifier] || [];
       this.pointsPerOption[identifier].push(response.responseDetails.answers[0]);
 
+      this.pointsPerOptionToSelf[identifier] = this.pointsPerOptionToSelf[identifier] || [];
+      if (response.giver === response.recipient) {
+        this.pointsPerOptionToSelf[identifier].push(response.responseDetails.answers[0]);
+      }
```
[the remark this answers is not in the document]
Comment (Jellybeano) [reply src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts:50]: fair point!

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts [hunk 7]
```diff
@@ -54,7 +63,16 @@ export class ConstsumRecipientsQuestionStatisticsCalculation
       const sum: number = answers.reduce((a: number, b: number) => a + b, 0);
       this.totalPointsPerOption[recipient] = sum;
       this.averagePointsPerOption[recipient] = +(answers.length === 0 ? 0 : sum / answers.length).toFixed(2);
+      if (this.pointsPerOptionToSelf[recipient].length === 0) {
+        this.averagePointsExcludingSelf[recipient] = sum;
```
[the remark this answers is not in the document]
Comment (Jellybeano) [reply src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts]: I cannot believe I somehow missed this out, thank you for catching this!

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts [hunk 8]
```diff
@@ -54,7 +63,10 @@ export class ConstsumRecipientsQuestionStatisticsCalculation
       const sum: number = answers.reduce((a: number, b: number) => a + b, 0);
       this.totalPointsPerOption[recipient] = sum;
       this.averagePointsPerOption[recipient] = +(answers.length === 0 ? 0 : sum / answers.length).toFixed(2);
+      this.averagePointsExcludingSelf[recipient] =
+          +(answers.length === 0 ? 0 : (sum - pointsPerOptionToSelf[recipient]) / answers.length).toFixed(2);
```
Comment (halfwhole) [inline src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts:67]: Slight issue with the following, didn't realise it myself until testing it either. Looking at Gene's line, the average excluding self ought to be 50, but it's 25, because it's incorrectly dividing by 2 instead of 1:

Screenshot 2021-06-30 at 1 27 50 AM

Will need to find a way to determine if `pointsPerOptionToSelf[recipient]` has actually been assigned a value or not. One way (if it's possible) is to allow it to be undefined instead of setting it to 0 by default above, then here use a conditional to check if it's been assigned a value -- then subtract and divide accordingly

Alternatively, could also directly capture the excluding-self-scores using `pointsPerOptionExcludingSelf`, `answersExcludingSelf`, `sumExcludingSelf` etc. instead of having to minus it off at the end

Comment (Jellybeano) [reply src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts]: This bug should hopefully be fixed now!

**Review by halfwhole — APPROVED** [review]
LGTM

Screenshot 2021-07-01 at 1 42 04 AM

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts [hunk 9]
```diff
@@ -54,7 +62,15 @@ export class ConstsumRecipientsQuestionStatisticsCalculation
       const sum: number = answers.reduce((a: number, b: number) => a + b, 0);
       this.totalPointsPerOption[recipient] = sum;
       this.averagePointsPerOption[recipient] = +(answers.length === 0 ? 0 : sum / answers.length).toFixed(2);
+      if (pointsPerOptionToSelf[recipient] === undefined) {
```
Comment (rrtheonlyone) [inline src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts:65]: better to do this like this:
```
this.averagePointsExcludingSelf[recipient] = this.averagePointsPerOption[recipient];
if (this.averagePointsPerOption[recipient] && pointsPerOptionToSelf[recipient] != undefined) {
     this.averagePointsExcludingSelf[recipient] = +(sum - pointsPerOptionToSelf[recipient]) / (answers.length - 1)).toFixed(2);
}
```

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Nice one!

One minor suggestion below. Let's also update `constsum-recipients-question-statistics.component.spec.ts` and add a little bit of tests to make sure we do not run into bugs down the road.

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts (hunk 9 above)
Comment (Jellybeano) [reply src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts]: thanks for the tip! the logic really does flow better this way

just to clarify, what exactly should be changed in `constsum-recipients-question-statistics.component.spec.ts`?

**Review by halfwhole — COMMENTED** [review]
The testing looks solid! We could add a few more cases to test though, especially boundary conditions (we'd like to hit all the conditions in lines 64-69 if possible):
- Average excluding self, when the recipient has self-review and is the only review
- Average excluding self, when the recipient has no self-review and is the only review
- Average excluding self, when the recipient has no self-review and is not the only review

File: src/web/app/components/question-types/question-statistics/constsum-recipients-question-statistics.component.spec.ts [hunk 10]
```diff
@@ -24,4 +27,145 @@ describe('ConstsumRecipientsQuestionStatisticsComponent', () => {
...
+      giver: 'Alice',
+      giverTeam: 'Team 1',
+      giverEmail: 'alice@gmail.com',
+      giverSection: '',
+      recipient: 'Bob',
+      recipientTeam: 'Team 2',
+      recipientEmail: 'bob@gmail.com',
+      recipientSection: '',
+      responseDetails: {
+        answers: [2, 8],
```
[the remark this answers is not in the document]
Comment (Jellybeano) [reply src/web/app/components/question-types/question-statistics/constsum-recipients-question-statistics.component.spec.ts]: Fair point!

File: src/web/app/components/question-types/question-statistics/constsum-recipients-question-statistics.component.spec.ts [hunk 11]
```diff
@@ -24,4 +27,145 @@ describe('ConstsumRecipientsQuestionStatisticsComponent', () => {
...
+      recipientEmail: 'emma@gmail.com',
+      recipientSection: '',
+      responseDetails: {
+        answers: [7, 3],
+        questionType: FeedbackQuestionType.CONSTSUM,
+      } as FeedbackConstantSumResponseDetails,
+    },
+  ];
+
+  it('should calculate Average Excluding Self correctly', () => {
```
[the remark this answers is not in the document]
Comment (Jellybeano) [reply src/web/app/components/question-types/question-statistics/constsum-recipients-question-statistics.component.spec.ts]: I named the test when I started and forgot to change it as I worked on the test case, thank you for pointing it out!

**Review by halfwhole — APPROVED** [review]

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts [hunk 12]
```diff
@@ -54,7 +62,13 @@ export class ConstsumRecipientsQuestionStatisticsCalculation
       const sum: number = answers.reduce((a: number, b: number) => a + b, 0);
       this.totalPointsPerOption[recipient] = sum;
       this.averagePointsPerOption[recipient] = +(answers.length === 0 ? 0 : sum / answers.length).toFixed(2);
+      this.averagePointsExcludingSelf[recipient] = this.averagePointsPerOption[recipient];
+      if (this.averagePointsPerOption[recipient] && pointsPerOptionToSelf[recipient] !== undefined) {
+        this.averagePointsExcludingSelf[recipient] =
```
Comment (rrtheonlyone) [inline src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts:67]: You do not need the ternary here since you already checked for `this.averagePointsPerOption[recipient]` not being zero

**Review by rrtheonlyone — APPROVED** [review]
Good to merge after a minor change below 👍

File: src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts (hunk 12 above)
Comment (Jellybeano) [reply src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts:67]: The ternary expression checks for edge cases where the recipient has only one response which is also a self response!

If the ternary expression is removed, situations such as below can appear:
image
because the only response received by Alice is a self-response

While I suppose such situations would be rare, I think it might be safer to leave the ternary expression there?
Or would it be better to shift the ternary expression outside of the if block?

Comment (rrtheonlyone) [reply src/web/app/components/question-types/question-statistics/question-statistics-calculation/constsum-recipients-question-statistics-calculation.ts:67]: Ah good catch! Let's leave it then :)
